packages/table/src/Row.tsx

```diff
@@ -0,0 +1,323 @@
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import { useTableContext, TableTypes, DataType } from './TableContext';
+import { tdInnerDiv } from './Cell';
+
+const rowStyle = css`
+  border-top: 1px solid ${uiColors.gray.light2};
+  color: ${uiColors.gray.dark2};
+
+  & > td > ${tdInnerDiv.selector} {
+    height: 40px;
+    overflow: hidden;
+    transition: all 150ms ease-in-out;
+  }
+`;
+
+const altColor = css`
+  &:nth-of-type(even) {
+    background-color: ${uiColors.gray.light3};
+  }
+`;
+
+const iconButtonMargin = css`
+  flex-grow: 1;
+  flex-shrink: 0;
+  margin-right: 4px;
+`;
+
+const disabledStyle = css`
+  background-color: ${uiColors.gray.light2};
+  color: ${uiColors.gray.base};
+  cursor: not-allowed;
+  border-top: 1px solid ${uiColors.gray.light1};
+  border-bottom: 1px solid ${uiColors.gray.light1};
+`;
+
+const displayFlex = css`
+  display: flex;
+  align-items: center;
+`;
+
+const truncation = css`
+  max-width: 100px;
+  white-space: nowrap;
+  text-overflow: ellipsis;
+  overflow: hidden;
+`;
+
+const transitionStyles = {
+  default: css`
+    transition: border 150ms ease-in-out;
+    border-top-color: transparent;
+
+    & > td {
+      padding-top: 0px;
+      padding-bottom: 0px;
+    }
+
+    & > td > ${tdInnerDiv.selector} {
+      max-height: 0;
+    }
+  `,
+
+  entered: css`
+    border-top-color: ${uiColors.gray.light2};
+
+    & > td > ${tdInnerDiv.selector} {
+      max-height: 40px;
+    }
+  `,
+};
+
+function styleColumn(index: string, dataType?: DataType) {
+  let justify;
+
+  if (dataType === DataType.Number) {
+    justify = 'flex-end';
+  } else {
+    justify = 'flex-start';
+  }
+
+  return css`
+    & td:nth-child(${index}) > div {
+      justify-content: ${justify};
+    }
+  `;
+}
+
+function getIndentLevelStyle(indentLevel: number) {
+  return css`
+    & > td:nth-child(1) {
+      padding-left: ${8 + indentLevel * 16}px;
+    }
+  `;
+}
+
+const idAllocator = IdAllocator.create('text-input');
+
+interface RowProps extends React.ComponentPropsWithRef<'tr'> {
+  expanded?: boolean;
+  disabled?: boolean;
+  indentLevel?: number;
+  isAnyAncestorCollapsed?: boolean;
+}
+
+const Row = React.forwardRef(
+  (
+    {
+      expanded = false,
+      disabled = false,
+      indentLevel = 0,
+      isAnyAncestorCollapsed: isAnyAncestorCollapsedProp,
+      children,
+      className,
+      ...rest
+    }: RowProps,
+    ref: React.Ref<any>,
+  ) => {
+    const {
+      state: { data, columnInfo, hasNestedRows, hasRowSpan },
+      dispatch: tableDispatch,
+    } = useTableContext();
+
+    const indexRef = useRef(idAllocator.generate());
+    const [isExpanded, setIsExpanded] = useState(expanded);
+    const nodeRef = useRef(null);
+
+    useEffect(() => {
+      let shouldDispatchHasNestedRows = false;
+      let shouldDispatchHasRowSpan = false;
+
+      React.Children.forEach(children, child => {
```

Comment:
If `hasNestedRows` and `hasRowSpan` are true we don't need to iterate, right?

Also if `data` changes we need to reset those values.

Reply:
true